### PR TITLE
fix(agents): AI_CHAT page scroll, reopen closed conversations, settings link

### PR DESCRIPTION
**Pull request by 2witstudios**
## Summary
- Fixes AI_CHAT pages (opened as a page, not in the console pane grid) being unscrollable — `CenterPanel` was double-wrapping chat's self-managed scroll region in its own `overflow-auto` `CustomScrollArea`. Added `managesOwnScroll` to `usePageRefresh()` (true for AI_CHAT/CHANNEL, and now also resolved for a page not yet in the tree via the same `/api/pages/${pageId}` fallback `CenterPanel` itself uses) and skip the redundant wrapper for those page types.
- Fixes the agents console permanently hiding a conversation once its pane is closed (`closedInSessionAt` excludes it from the sidebar with no way back). Adds a symmetric reopen path (`reopen-conversation-in-session.ts`, mirroring the existing close decision's cap/lock discipline), `GET .../conversations/closed` + `POST .../conversations/[id]/reopen` routes, and a History section in `AgentsSidebar` to list and reopen them.
- Adds a Settings entry point to the console — a link on each pane that opens the agent's Settings tab (`/p/[pageId]?tab=settings`, which `AgentPageView` now reads to land directly on Settings, and re-syncs if the same instance receives the param again via a query-only navigation).

## Review history
This PR went through six rounds of automated review (chatgpt-codex-connector, coderabbitai) plus independent adversarial passes — every finding was verified against current code and either fixed or confirmed already-handled with a regression test proving it:
- **Race condition**: `createConversationInSession` wasn't under the same per-session advisory lock as close/reopen, so a create racing a reopen for a session's last open slot could both succeed, exceeding `MAX_SESSION_CONVERSATIONS`. Fixed, with a real-Postgres concurrency test (`agent-sessions-runtime.integration.test.ts`) firing both simultaneously.
- **Tab-sync bug**: the pane bar's Settings link only worked on a fresh mount; a query-only navigation to an already-mounted `AgentPageView` silently no-opped. Fixed with a reactive effect + regression tests, including the exact "identical value re-navigated" edge case CodeRabbit raised.
- **Scroll-fix gap**: `usePageRefresh` didn't account for a page CenterPanel renders via its tree-fallback fetch (a freshly-created or deep-linked page). Fixed by mirroring that same fallback fetch.
- **NULL-sort bug**: the closed-conversations list ordered by `desc(lastMessageAt)`, but Postgres sorts NULLs first — a never-messaged closed conversation could crowd out real activity from the capped History list. Fixed with `NULLS LAST`.
- **Non-atomic soft-delete**: `softDeleteConversation` only deactivated `chatMessages`, leaving `conversations.isActive` stale, and its two updates weren't transactional. Fixed — both updates now run in one `db.transaction`, and the conversation row is deactivated too.
- **History-emptying a session**: deleting a conversation from History bypassed the never-empty guard `closeConversationInSession` enforces, so History could empty an active session down to zero reachable conversations. Fixed with the same guard (409 + `reason: 'last_conversation'`) on the DELETE route.
- **Drive-scoped cache miss / duplicate-insert on reopen**: reopening from the drive-scoped view (`/dashboard/agents`) only updated the session-scoped SWR cache, and calling both the session- and drive-scoped updaters could double-insert the same conversation. Fixed with a scoped `mutate` call and an idempotent (de-duping) cache updater.
- **History going stale after a pane close**: closing a pane didn't refresh an already-open History list, so a just-closed conversation wouldn't appear there until an unrelated refresh. Fixed.
- **Pool-starvation deadlock (P1)**: the session-listing lock used `db.transaction()` + `pg_advisory_xact_lock`, pinning the main pool's connection for the transaction's duration while its own callback (via `conversationRepository`/global `db`) needed a *second* connection from that same pool — a real deadlock under `DB_POOL_MAX=1` or a pool-size burst. Replaced with `withSessionListingLock`, built on this codebase's existing `withAdvisoryLock` primitive against the dedicated `getAdvisoryLockPool()` (structurally separate from the main pool), with bounded retry-then-throw instead of silent degrade-to-unlocked. Applied across create/close/reopen.
- **Unserialized History-deletion guard (P2)**: the never-empty guard's count-then-delete used the explicitly lock-free `countOpenConversationsForSession`, so two racing History deletes (or one racing the normal close route) could both read a stale open count and both empty the session. Fixed by running the guard inside the same `withSessionListingLock` create/close/reopen already serialize under.
- Plus: DRY'd the duplicated close/reopen read-deps, tightened a type in the query-forwarding redirect, and extended sidebar tests to assert the actual toast + a dedicated 404 case.

Merged with master's PR #2297 (`pu/session-refresh-persist`, an unrelated tab-history/back-forward fix) — zero textual conflicts (`git merge-tree` confirmed clean before merging), re-verified with a full local validation pass afterward.

## Test plan
- [x] `bunx tsc --noEmit` clean across the whole app
- [x] `eslint` clean on all changed files (only pre-existing warnings in untouched files)
- [x] 798 unit tests passing across the affected suites (agents console, agent-sessions API + lib, usePageRefresh, conversation-repository)
- [x] 2 real-Postgres integration tests passing (`scripts/test-with-db.sh` DB): the cap-race concurrency test (re-verified against the new lock mechanism) and the NULLS-LAST ordering test
- [x] All CI checks green (Lint & TypeScript, Unit Tests, Security Test Suite, Dependency Audit, CodeQL, Static Security Analysis, Secret Scanning)
- [ ] Manual click-through in a live browser (not completed in this session — no seeded admin credentials / DB hostname reachable from a host-run dev server in this worktree; recommend a quick smoke test before merge)

🤖 Generated with [Claude Code](https://claude.com/claude-code)

https://claude.ai/code/session_01PofpNe4dkeFdFrN3TjhVFd

## Summary by CodeRabbit

- **New Features**
  - Added History views for agent sessions, including lazy loading, empty/error states, retry support, and reopening closed conversations.
  - Added direct links to agent settings from conversation panes.
  - Added URL support for opening agent pages on Chat, History, or Settings tabs.
  - Preserved query parameters during page redirects and authentication flows.

- **Improvements**
  - Improved scrolling behavior on chat and channel pages.
  - Reopening conversations now refreshes lists and selects the restored conversation automatically.
  - Prevented deletion of the last active conversation in a session.
  - Improved consistency when conversations are deleted or reopened concurrently.

### CODE REVIEW

File: apps/web/src/components/agents/AgentPageView.tsx [hunk 1]
```diff
@@ -107,7 +108,15 @@ export default function AgentPageView({ page }: AgentPageViewProps) {
...
-  const [activeTab, setActiveTab] = useState<string>('chat');
+  // `?tab=settings` deep-links here (the agents console's Settings link, via
+  // `/p/[pageId]` — that redirect route forwards the query string verbatim).
+  // Read once at mount, not synced afterward: this page's own tab clicks are
+  // the source of truth from then on, and `useSearchParams` re-rendering the
+  // initializer would fight a user who has since clicked a different tab.
+  const initialTab = useSearchParams().get('tab');
+  const [activeTab, setActiveTab] = useState<string>(
+    initialTab === 'history' || initialTab === 'settings' ? initialTab : 'chat',
+  );
```
[the remark this answers is not in the document]
Comment (2witstudios) [reply apps/web/src/components/agents/AgentPageView.tsx]: Fixed in 07e96b3c7: activeTab was only seeded from `?tab=` at mount via a `useState` initializer, so a same-instance query-only navigation (this pane pointing at its own page, the common case) silently no-opped as you described. Added a reactive `useEffect` keyed on `searchParams` that re-syncs `activeTab` (and strips the param) on every change, not just at mount. Added 4 regression tests in `AgentPageView.test.tsx` covering fresh-mount, re-sync-while-mounted, param-stripping, and an unrecognized-value no-op.

File: apps/web/src/hooks/usePageRefresh.ts [hunk 2]
```diff
@@ -43,8 +51,14 @@ export function usePageRefresh(): PageRefreshConfig {
   // Create a stable no-op function
   const noOp = useCallback(async () => {}, []);
 
+  // Chat-shaped pages own their scroll end to end; see `managesOwnScroll` on
+  // `PageRefreshConfig`. Computed once here so it can be merged onto every
+  // branch below without repeating the check in each one.
+  const managesOwnScroll =
+    currentPage?.type === PageType.AI_CHAT || currentPage?.type === PageType.CHANNEL;
```
[the remark this answers is not in the document]
Comment (2witstudios) [reply apps/web/src/hooks/usePageRefresh.ts:82]: Fixed in 07e96b3c7: `usePageRefresh` now mirrors `CenterPanel`/`PageContent`'s own `/api/pages/${pageId}` fallback fetch (same gate: tree settled, pageId present, not already found), so `managesOwnScroll` agrees with what's actually on screen even before the page lands in the tree. Added `usePageRefresh.test.ts` covering the in-tree case, the fallback-resolves case, and that the fallback fetch is correctly gated (not fired while the tree is still loading, or with no pageId).

File: apps/web/src/lib/agent-sessions/agent-sessions-runtime.ts [hunk 3]
```diff
@@ -334,6 +338,101 @@ export async function closeConversationInSession(input: {
...
+ * of the same session's listings.
+ */
+export async function reopenConversationInSession(input: {
+  conversationId: string;
+  sessionId: string;
+}): Promise<ReopenConversationOutcome> {
+  return db.transaction(async (tx) => {
+    await tx.execute(
+      sql`SELECT pg_advisory_xact_lock(hashtext(${'agent-session-conversations:' + input.sessionId}))`,
+    );
```
[the remark this answers is not in the document]
Comment (2witstudios) [reply apps/web/src/lib/agent-sessions/agent-sessions-runtime.ts]: Fixed in 07e96b3c7: wrapped `createConversationInSession` in the same per-session advisory lock (`db.transaction` + `pg_advisory_xact_lock`, same key) that close/reopen already use, so all three operations that touch the open-listing cap now serialize against each other. Added `agent-sessions-runtime.integration.test.ts`, a real-Postgres concurrency test that fires a create and a reopen at a session's last open slot simultaneously and asserts the final row count never exceeds `MAX_SESSION_CONVERSATIONS`. Worth noting: writing that test caught a bug in the test's own assertion logic first (a fulfilled `'session_full'` reopen was being double-counted as both succeeded and refused) — fixed before trusting the result, and the fix is confirmed correct against genuine concurrent transactions, not mocks.

File: apps/web/src/components/agents/AgentPageView.tsx [hunk 4]
```diff
@@ -138,7 +108,15 @@ export default function AgentPageView({ page }: AgentPageViewProps) {
...
```
[the remark this answers is not in the document]
Comment (2witstudios) [reply apps/web/src/components/agents/AgentPageView.tsx]: Verified against the current code rather than just asserted: the fix from 07e96b3c7 already handles this. The effect does not diff against a previously-seen value — it applies whatever valid `tab` is present on every `searchParams` identity change, and Next produces a new `ReadonlyURLSearchParams` instance on every real navigation (including one that lands on the same string as before). Added a regression test in `AgentPageView.test.tsx` (`re-applies a repeat ?tab=settings navigation even after the user has since clicked away`) that clicks Settings, clicks Chat, then clicks Settings again via a second independent `searchParams` change — it lands on Settings correctly with no previous-query tracking needed. Went with the simpler unconditional-apply design over the suggested previous-query-detector since it already covers the identical-value case by construction, and adding tracked state for it would be unnecessary complexity with no behavioral gain.

File: apps/web/src/lib/agent-sessions/agent-sessions-runtime.ts [hunk 5]
```diff
@@ -334,6 +338,101 @@ export async function closeConversationInSession(input: {
...
+    .from(conversations)
+    .where(
+      and(
+        eq(conversations.sessionId, sessionId),
+        eq(conversations.isActive, true),
+        isNotNull(conversations.closedInSessionAt),
+      ),
+    )
+    .orderBy(desc(conversations.lastMessageAt))
+    .limit(MAX_SESSION_CONVERSATIONS);
```
[the remark this answers is not in the document]
Comment (2witstudios) [reply apps/web/src/lib/agent-sessions/agent-sessions-runtime.ts]: Fixed in 452ed5d5b: applied your suggested `DESC NULLS LAST` fix exactly. Added a real-Postgres test (`agent-sessions-runtime.integration.test.ts`) with a never-messaged closed conversation alongside older/newer real-activity ones — Postgres NULL-sort defaults aren't something a mock can verify, so this needed the real DB the rest of this file's integration tests already use.

File: apps/web/src/lib/agent-sessions/agent-sessions-runtime.ts [hunk 6]
```diff
@@ -334,6 +373,82 @@ export async function closeConversationInSession(input: {
...
+      title: conversations.title,
+      type: conversations.type,
+      contextId: conversations.contextId,
+      lastMessageAt: conversations.lastMessageAt,
+    })
+    .from(conversations)
+    .where(
+      and(
+        eq(conversations.sessionId, sessionId),
+        eq(conversations.isActive, true),
```
[the remark this answers is not in the document]
Comment (2witstudios) [reply apps/web/src/lib/agent-sessions/agent-sessions-runtime.ts:479]: Confirmed and fixed in f379d0055. Investigated the blast radius first (only one call site of `softDeleteConversation`; no test relies on the old behavior; `conversations.isActive`'s own schema doc already declares its intended meaning as "history soft-delete"; the sibling `globalConversationRepository.softDeleteConversation` already does this correctly) before changing a widely-shared repository function. Now sets `conversations.isActive = false` for that conversationId alongside the existing `chatMessages.isActive = false` update. This also closes two other pre-existing gaps the same investigation surfaced: the compliance retention purge never reclaimed history-deleted page conversations, and the v1/MCP conversations API was leaking them. Added a mocked-DB repository test asserting both UPDATE statements fire against the right tables.

File: apps/web/src/components/layout/left-sidebar/AgentsSidebar.tsx [hunk 7]
```diff
@@ -625,6 +647,49 @@ function SessionRow({
...
+        await post(
+          `/api/agent-sessions/${encodeURIComponent(session.sessionId)}/conversations/${encodeURIComponent(conversation.conversationId)}/reopen`,
+        );
+        // Drop it from the closed list and refresh the open list — it now
+        // belongs in both places (the second matters even before the poll
+        // catches up, or the row a user just reopened looks like a no-op).
+        void retryClosed();
+        onChanged();
+        setExpanded(true);
+        openConversation(conversation);
```
[the remark this answers is not in the document]
Comment (2witstudios) [reply apps/web/src/components/layout/left-sidebar/AgentsSidebar.tsx:744]: Confirmed and fixed in f379d0055 — went with your "optimistically add it to the agent-sessions cache" option. Verified `agentSessionsKey(driveId)` (AgentPanes' cache key) and this sidebar's own `sessionsKey` are the identical string, so it's the same SWR cache entry, not a driveId-scoped mismatch. Extended `onChanged` to optionally take an updater applied via `mutate(updater, {revalidate: false})` — the exact pattern `AgentPanes.tsx`'s own `recordMintedConversation`/`recordClosedConversation` already use on this same cache — and call it with the reopened conversation prepended to its session's list, synchronously, before `openConversation` focuses the pane. Added a test proving the open-conversations list shows the reopened entry with zero additional sessions-list fetches.

File: apps/web/src/components/layout/left-sidebar/AgentsSidebar.tsx [hunk 8]
```diff
@@ -625,6 +666,67 @@ function SessionRow({
...
+        // chatgpt-codex-connector on PR #2296). Mirrors `AgentPanes.tsx`'s
+        // own `recordMintedConversation`/`recordClosedConversation`.
+        void retryClosed();
+        onChanged((current) => {
+          if (!current) return current;
+          return {
+            sessions: current.sessions.map((s) =>
+              s.sessionId === session.sessionId
+                ? { ...s, conversations: [conversation, ...s.conversations] }
+                : s,
```
[the remark this answers is not in the document]
Comment (2witstudios) [reply apps/web/src/components/layout/left-sidebar/AgentsSidebar.tsx]: Confirmed and fixed in 825c75902. Now also writes directly to `agentSessionsKey(session.driveId)` (the exact key `AgentPanes` reads) via `useSWRConfig()`'s scoped `mutate` — this codebase's own established pattern for this (`usePageTree.ts`, `QuickCreatePalette.tsx`, etc. already use it), and correct where a bare `mutate` import from `'swr'` would only ever target the true default cache. Safe to call unconditionally alongside the existing sidebar-key write, since in drive-scoped view the two keys already coincide.

That coincidence surfaced a real bug while writing the regression test: when both calls land on the same cache entry, a plain prepend applied twice inserted the conversation twice (an actual React duplicate-key warning, not hypothetical) — fixed by making the insert idempotent (de-dupe by conversationId first). Verified with a probe component mounted in the same `SWRConfig` as the sidebar, reading the exact key `AgentPanes` reads, in both the global (non-coinciding) and drive-scoped (coinciding) cases.

File: apps/web/src/lib/repositories/conversation-repository.ts [hunk 9]
```diff
@@ -375,6 +375,20 @@ export const conversationRepository = {
...
+    // (agent-sessions-runtime.ts's session listings/caps, the v1/MCP
+    // conversations API, the compliance retention purge) previously kept
+    // treating a page conversation deleted from History as live forever —
+    // including, most concretely, letting the agents console's reopen
+    // affordance resurrect one with no messages left (review finding —
+    // chatgpt-codex-connector on PR #2296).
+    await db
+      .update(conversations)
+      .set({ isActive: false })
+      .where(eq(conversations.id, conversationId));
```
[the remark this answers is not in the document]
Comment (2witstudios) [reply apps/web/src/lib/repositories/conversation-repository.ts]: Confirmed and fixed in 825c75902 — wrapped both UPDATEs in a single `db.transaction`. Added a rollback test asserting the `conversations` UPDATE never runs at all if the `chat_messages` UPDATE throws (matching real transaction semantics: the callback aborts before its next statement).

File: apps/web/src/lib/repositories/conversation-repository.ts [hunk 10]
```diff
@@ -368,13 +368,34 @@ export const conversationRepository = {
...
+          eq(chatMessages.conversationId, conversationId)
+        ));
+      // The canonical row itself, not just its messages — mirrors
+      // `globalConversationRepository.softDeleteConversation`'s pattern, and
+      // matches `conversations.isActive`'s own doc comment ("history soft-
+      // delete"). Every reader that gates on `conversations.isActive`
+      // (agent-sessions-runtime.ts's session listings/caps, the v1/MCP
+      // conversations API, the compliance retention purge) previously kept
+      // treating a page conversation deleted from History as live forever —
+      // including, most concretely, letting the agents console's reopen
```
[the remark this answers is not in the document]
Comment (2witstudios) [reply apps/web/src/lib/repositories/conversation-repository.ts:392]: Confirmed and fixed in 3f4a96cb0. Added the same never-empty guard `closeConversationInSessionWith` already enforces to the page-agent conversations DELETE route: when the conversation is session-bound, still open (`closedInSessionAt` null), and the session's last one (via the existing `countOpenConversationsForSession` reader), refuse with 409 (`reason: 'last_conversation'`, same shape as the close route's own refusal) instead of letting the delete silently empty the session. Went with a route-level guard rather than teaching `softDeleteConversation` about session lifecycle, matching this codebase's own architecture (lifecycle decisions live in dedicated modules/call sites, not the repository layer). Added 4 tests: refuses the last-listing case, allows deletion when another listing remains, and confirms the guard is correctly skipped both for already-closed and non-session-bound conversations.

File: apps/web/src/components/layout/left-sidebar/AgentsSidebar.tsx [hunk 11]
```diff
@@ -526,20 +552,41 @@ function SessionRow({
...
+  // since most sessions are browsed without ever needing it.
+  const {
+    data: closedData,
+    error: closedError,
+    isLoading: closedLoading,
+    mutate: retryClosed,
+  } = useSWR<{ conversations: SessionConversationEntry[] }>(
+    showHistory ? `/api/agent-sessions/${encodeURIComponent(session.sessionId)}/conversations/closed` : null,
+    closedConversationsFetcher,
+    { revalidateOnFocus: false },
```
[the remark this answers is not in the document]
Comment (2witstudios) [reply apps/web/src/components/layout/left-sidebar/AgentsSidebar.tsx]: Confirmed and fixed in 3f4a96cb0, two ways: `closeConversation` now also calls `retryClosed()` directly (immediate feedback when closed from this row's own visible open list), and the closed-conversations fetch now polls every 20s like the main sessions list already does (catches every other source, e.g. a pane closed from `AgentPanes` elsewhere, without needing cross-component plumbing). Added tests for both: the already-open-History case refetches, and closing while History is collapsed is a safe no-op (its SWR key is null then).

File: apps/web/src/lib/agent-sessions/agent-sessions-runtime.ts [hunk 12]
```diff
@@ -230,50 +247,96 @@ export async function createConversationInSession(input: {
...
-          );
-        return row?.n ?? 0;
+  return db.transaction(async (tx) => {
+    await tx.execute(
+      sql`SELECT pg_advisory_xact_lock(hashtext(${'agent-session-conversations:' + input.sessionId}))`,
+    );
+    return createConversationInSessionWith(
+      {
+        createPageConversation: ({ conversationId, userId, agentPageId, sessionId, title }) =>
+          conversationRepository.createConversation(conversationId, userId, agentPageId, {
```
[the remark this answers is not in the document]
Comment (2witstudios) [reply apps/web/src/lib/agent-sessions/agent-sessions-runtime.ts:313]: Fixed — but via a different mechanism than passing `tx` through, since that approach reintroduces the exact hazard this thread is about: `db.transaction()` pins one connection for the whole callback, and anything the callback calls that still reaches for the plain `db` (or a nested `db.transaction()`) needs a *second* connection from that same pool while the first is held — which is the `DB_POOL_MAX=1` deadlock this finding describes, just one layer down (e.g. `conversationRepository.getConversation`/`getAiAgent` and `resolveOrCreateConversation` all use plain `db` internally, and threading `tx` through every one of those call sites would require rewriting a large surface of shared repository code to be transaction-aware end-to-end).

Instead, `createConversationInSession`/`closeConversationInSession`/`reopenConversationInSession` now go through a new `withSessionListingLock(sessionId, fn)` (agent-sessions-runtime.ts) built on this codebase's existing `withAdvisoryLock` primitive (`@pagespace/db/advisory-lock`) against `getAdvisoryLockPool()` — a dedicated pool completely separate from the main `db` pool, used the same way `start-generation-exclusive.ts` already isolates its own advisory lock from main-pool contention. The mutual-exclusion guarantee is unchanged (still a per-session advisory lock serializing create/close/reopen), but the lock connection and the query connections are now structurally independent, so the callback's queries against plain `db` can never starve waiting on the very connection holding their own lock — including under `DB_POOL_MAX=1`. Bounded retry-then-throw (`SESSION_LISTING_LOCK_MAX_RETRIES`/`SessionListingLockBusyError`) replaces the old `pg_advisory_xact_lock`, so lock contention fails loudly instead of silently degrading unlocked.

Verified via the existing real-Postgres concurrency test (`agent-sessions-runtime.integration.test.ts`, create-vs-reopen racing for a session's last slot) passing unchanged against the new locking mechanism.

File: apps/web/src/app/api/ai/page-agents/[agentId]/conversations/[conversationId]/route.ts [hunk 13]
```diff
@@ -218,6 +219,37 @@ export async function DELETE(
...
+    // `closeConversationInSessionWith`'s never-empty guard. Refusing here
+    // when this is the session's LAST open listing keeps that same
+    // invariant true regardless of which route emptied it (review finding —
+    // chatgpt-codex-connector on PR #2296): an active session left with a
+    // live sandbox and zero reachable conversations is worse than a plain
+    // 409 telling the user to close the pane (which offers to end the
+    // session) or delete history from a different conversation first.
+    if (conversationRow?.sessionId && conversationRow.closedInSessionAt === null) {
+      const openCount = await countOpenConversationsForSession(conversationRow.sessionId);
+      if (openCount <= 1) {
```
[the remark this answers is not in the document]
Comment (2witstudios) [reply apps/web/src/app/api/ai/page-agents/[agentId]/conversations/[conversationId]/route.ts]: Fixed — the DELETE route's never-empty guard (count-then-soft-delete) now runs inside `withSessionListingLock(sessionId, fn)`, the same per-session advisory lock `createConversationInSession`/`closeConversationInSession`/`reopenConversationInSession` already serialize under (see the reply on the sibling P1 thread for how that lock is now built). `countOpenConversationsForSession` and `conversationRepository.softDeleteConversation` both execute inside that lock's callback, so two racing History DELETEs — or one racing the normal close route — can no longer both observe count=2 and independently empty the session; the second caller now waits for the first to finish and re-reads a consistent count. `withSessionListingLock` is exported from `agent-sessions-runtime.ts` specifically for this route to reuse rather than duplicating lock logic.

Covered by new tests in the route's test file asserting `withSessionListingLock` is called with the session ID around the guard-plus-delete, and by the existing real-Postgres concurrency test confirming the shared lock mechanism still correctly serializes concurrent callers.

File: apps/web/src/app/api/ai/page-agents/[agentId]/conversations/[conversationId]/route.ts [hunk 14]
```diff
@@ -221,8 +222,51 @@ export async function DELETE(
...
+            reason: 'last_conversation',
+          },
+          { status: 409 },
+        );
+      }
+    } else {
+      // Not session-bound (or already closed out of its session, so it
+      // holds no open slot to protect) — no lock needed, nothing to
+      // serialize against.
+      await conversationRepository.softDeleteConversation(agentId, conversationId);
```
[the remark this answers is not in the document]
Comment (2witstudios) [reply apps/web/src/app/api/ai/page-agents/[agentId]/conversations/[conversationId]/route.ts:297]: Fixed in 3fd45bfbb — every session-bound deletion now takes `withSessionListingLock`, regardless of the row's open/closed state at the pre-lock read, and re-reads the row's live state once actually inside the lock rather than branching on the stale pre-lock snapshot (which was only ever used for the initial ownership/audit reads, not the lifecycle decision).

This closes both gaps your report identified: a closed-conversation delete now serializes against a concurrent reopen through the same per-session lock (so the interleaving your report describes — reopen reads `isActive: true`, unlocked delete commits `isActive: false`, reopen's `closedInSessionAt IS NOT NULL`-only guard still fires — can no longer happen, since both now execute one-at-a-time under the same lock key), and as a side effect the never-empty guard itself is now immune to a stale pre-lock snapshot: a conversation that looked closed at the initial read but was reopened by a concurrent request before the lock was acquired is re-read fresh inside the lock and correctly weighed against the cap. A row already inactive by the time the lock is acquired (e.g. a second delete racing the first) is treated as an idempotent no-op.

3 new regression tests cover: the closed-conversation-still-takes-the-lock case, the stale-snapshot-vs-fresh-read guard case (closed at outer read, open by the time the lock's fresh read runs — still correctly refused as the last open listing), and the idempotent-double-delete case.

File: apps/web/src/app/api/ai/page-agents/[agentId]/conversations/[conversationId]/route.ts [hunk 15]
```diff
@@ -221,8 +222,73 @@ export async function DELETE(
...
+          return 'already_deleted' as const;
+        }
+        if (fresh.closedInSessionAt === null) {
+          // Still open — occupies a listing slot, so the never-empty guard applies.
+          const openCount = await countOpenConversationsForSession(sessionId);
+          if (openCount <= 1) return 'last_conversation' as const;
+        }
+        await conversationRepository.softDeleteConversation(agentId, conversationId);
+        return 'deleted' as const;
+      });
```
[the remark this answers is not in the document]
Comment (2witstudios) [reply apps/web/src/app/api/ai/page-agents/[agentId]/conversations/[conversationId]/route.ts:273]: Fixed in 417bd9dfe, essentially your suggested diff — `!fresh || !fresh.isActive` is now the only `already_deleted` case, and the never-empty guard is skipped (but the delete itself is NOT) when `fresh.sessionId !== sessionId`. New regression test: a row detached mid-lock (session cleared between the pre-lock and lock-held reads) still gets soft-deleted.

File: apps/web/src/lib/agent-sessions/agent-sessions-runtime.ts [hunk 16]
```diff
@@ -330,8 +421,81 @@ export async function closeConversationInSession(input: {
...
+}): Promise<ReopenConversationOutcome> {
+  return withSessionListingLock(input.sessionId, () =>
+    reopenConversationInSessionWith(
+      {
+        ...sessionListingReadDeps(),
+        reopenConversation: async (conversationId) => {
+          const updated = await db
+            .update(conversations)
+            .set({ closedInSessionAt: null })
+            .where(and(eq(conversations.id, conversationId), isNotNull(conversations.closedInSessionAt)))
```
[the remark this answers is not in the document]
Comment (2witstudios) [reply apps/web/src/lib/agent-sessions/agent-sessions-runtime.ts:447]: Fixed in 417bd9dfe — the global-assistant conversation DELETE route (`/api/ai/global/[id]`) now takes the identical `withSessionListingLock` treatment the page-agent route got in the previous round: every session-bound deletion acquires the lock, re-reads the row fresh inside it, and weighs the never-empty guard when still bound to that session. `globalConversationRepository`'s `Conversation` return type is widened with `sessionId`/`closedInSessionAt` (the query already selected them; only the type omitted them). 6 new regression tests cover the never-empty guard, the lock-on-a-closed-row case, the detached-but-active case, and idempotent double-delete — mirroring the page-agent route's own coverage.

File: apps/web/src/components/layout/left-sidebar/__tests__/AgentsSidebar.test.tsx [hunk 17]
```diff
@@ -676,6 +746,160 @@ describe('AgentsSidebar', () => {
...
+      await user.click(await screen.findByText('Researcher — Old chat'));
+      await waitFor(() => expect(mockPost).toHaveBeenCalled());
+
+      // Now showing in the (still-expanded) open conversations section too —
+      // present without any additional sessions-list fetch.
+      expect(screen.getAllByText('Researcher — Old chat').length).toBeGreaterThan(0);
+      const sessionsListCallsAfter = mockFetchWithAuth.mock.calls.filter(
+        ([url]) => url === '/api/agent-sessions?driveId=drive-1',
+      ).length;
+      expect(sessionsListCallsAfter).toBe(sessionsListCallsBefore);
```
[the remark this answers is not in the document]
Comment (2witstudios) [reply apps/web/src/components/layout/left-sidebar/__tests__/AgentsSidebar.test.tsx:837]: Fixed in 417bd9dfe, per your suggested diff — captures the label count before the reopen click and asserts it strictly increases, rather than just asserting presence (which the still-expanded History row alone would already satisfy).

File: apps/web/src/components/layout/left-sidebar/__tests__/AgentsSidebar.test.tsx [hunk 18]
```diff
@@ -676,6 +746,160 @@ describe('AgentsSidebar', () => {
...
+      renderSidebar();
+
+      await user.click(await screen.findByLabelText(/expand api refactor/i));
+      await user.click(await screen.findByText('History'));
+      const fetchesBeforeClick = mockFetchWithAuth.mock.calls.length;
+      await user.click(await screen.findByText('Researcher — Old chat'));
+
+      await waitFor(() => expect(mockPost).toHaveBeenCalled());
+      // `retryClosed()` re-fetches the closed list so the stale row can drop out.
+      await waitFor(() => expect(mockFetchWithAuth.mock.calls.length).toBeGreaterThan(fetchesBeforeClick));
```
[the remark this answers is not in the document]
Comment (2witstudios) [reply apps/web/src/components/layout/left-sidebar/__tests__/AgentsSidebar.test.tsx]: Fixed in 417bd9dfe, per your suggested diff — filters `mockFetchWithAuth` calls to the closed-list URL specifically (matching the close-action test's own pattern a few lines up) instead of counting any fetch.

File: apps/web/src/lib/agent-sessions/__tests__/agent-sessions-runtime.integration.test.ts [hunk 19]
```diff
@@ -0,0 +1,186 @@
...
+    expect(refused).toHaveLength(1);
+
+    // The refused side must be refused for the RIGHT reason (the cap), not
+    // some unrelated failure masking a real bug.
+    if (createResult.status === 'rejected') {
+      expect(createResult.reason).toBeInstanceOf(SessionFullError);
+    }
+    if (reopenResult.status === 'fulfilled') {
+      expect(reopenResult.value === 'reopened' || reopenResult.value === 'session_full').toBe(true);
+    }
```
[the remark this answers is not in the document]
Comment (2witstudios) [reply apps/web/src/lib/agent-sessions/__tests__/agent-sessions-runtime.integration.test.ts:137]: Fixed in 417bd9dfe, per your suggested diff — added `expect(reopenResult.status).toBe('fulfilled')` so a rejection on the reopen side (e.g. a lock failure) can no longer silently pass as a legitimate cap refusal.
